ImpPager.class.nut

```diff
@@ -36,78 +47,152 @@ class ImpPager {
+                        next(false);
+                        return;
+                    }
+
+                    // Save SPI Flash address in the message metadata
+                    if(!("metadata" in dataPoint)) dataPoint.metadata <- {}
+                    dataPoint.metadata.addr <- addr;
+
+                    if("rtc" in dataPoint.metadata && dataPoint.metadata.rtc == false){
+                      if(_lastTS == null){
+                        _lastTS = [hardware.millis(), time()] //With these two datapoints, we can now re-establish all of our timestamps
+                        _log_debug("Discovered most recent datapoint saved to SPIFlash without RTC - " + dataPoint.id + " attempting to rebuild timestamps with ms = " + _lastTS[0] + " and time = " + _lastTS[1])
+                      }
+
+                      _log_debug("Found log without RTC. ID=" + dataPoint.id + " and ts=" +dataPoint.ts)
+
+                      if("boot" in dataPoint.metadata && dataPoint.metadata.boot == _bootNumber){
+                        local deltaTMillis = _lastTS[0] - dataPoint.ts
+                        local deltaTSeconds = math.floor(deltaTMillis+500).tointeger()/1000 //Round to nearest second, but use this int value for updating _lastTS to keep things consistent
+                        dataPoint.ts = _lastTS[1] - deltaTSeconds  //All integer math, so no need to worry about decimal points
+
+                        _log_debug("Calculated new ts as " + dataPoint.ts + " (deltaT = " + deltaTMillis + " ms)")
+
+                        //update _lastTS so that we can have 25 days between datapoints instead of 25 days total of timestamps that we can rebuild
+                        _lastTS[0] -= (deltaTSeconds*1000)
+                        _lastTS[1] = dataPoint.ts
+
+                        // Our RTC has been reset - delete the metadata
+                        delete dataPoint.metadata.rtc
+
+                        // Update the dataPoint in our SPIFlash with its proper RTC.
+                        local newAddr = _spiFlashLogger.getPosition() + _spiFlashLogger._start;
+                        delete dataPoint.metadata.addr
+                        _spiFlashLogger.write(dataPoint)
+                        dataPoint.metadata.addr <- newAddr
+                        _log_debug("Wrote data with updated RTC to addr " + newAddr + " and Deleting datapoint at addr " + addr)
+
+                        // Delete the old dataPoint without the RTC.
+                        _spiFlashLogger.erase(addr);
+
+                      } else {
+                        server.error("Warning - dataPoint bootNumber " + dpBootNum + " != device bootNumber " + _bootNumber + " for message ID " + dataPoint.id + ".  Unable to rebuild ts...")
+                      }
+                    }
+
+
+                    _resendLoggedData(dataPoint);
+
+                    // Don't do any further scanning until we get an ACK for already sent message
+                    _next = next
```

Comment:
To reduce the risk of a race condition we probably need to have this statement at the beginning of the function.